### ISSUE TITLE
Expose additional IDs in quality forms

### ISSUE THREAD

**Issue by louisdesilva** [MyPureCloud/terraform-provider-genesyscloud#1066]
I will appreciate if you could expose questionGroups, questions and answerOptions ID to genesyscloud provider in terraform.

From the API explorer, the data is there but not available within terraform provider.

Thank you.

**Comment by charliecon** [comment]
Hi @louisdesilva 

Could you send URL of the endpoint on the API Explorer that has these fields, and also let me know what terraform resource specifically is missing them? I can see that these fields already exist inside "genesyscloud_quality_forms_evaluation" 

Thanks,
Charlie

**Comment by louisdesilva** [comment]
Hi Charlie,

As per https://developer.genesys.cloud/devapps/api-explorer#put-api-v2-quality-surveys-scorable, in the "answer" section, you can see that it only is accepting "questionGroupId", "questionId" and "answerId".

Where the "genesyscloud_quality_forms_evaluation" does not provides those Ids.

Regards,
Louis



General

________________________________

**Comment by charliecon** [comment]
Just so I understand better, does [PUT /api/v2/quality/surveys/scorable](https://developer.genesys.cloud/devapps/api-explorer#put-api-v2-quality-surveys-scorable) update an evaluation form? You can see [here](https://registry.terraform.io/providers/MyPureCloud/genesyscloud/latest/docs/resources/quality_forms_evaluation#nestedblock--question_groups--questions--answer_options) on our Terraform registry the endpoints that the evaluation forms resource uses. 

If there is a new field in the request body for [POST /api/v2/quality/forms/evaluations](https://developer.genesys.cloud/devapps/api-explorer#post-api-v2-quality-forms-evaluations) that is not in the resource, it will be a simple change for us to make. If we will need to implement a new endpoint into the resource, it will take more time.

**Comment by louisdesilva** [comment]
Yes, I am working on Terraform that deployed the form and data actions to update the form at the end of the survey using a workflow.

To work around now, I ran terraform to deploy the survey form, then I query the formID via the API to obtain the questionsID and answersID, I then put them into a varset and ran terraform again to deploy the data-action with the IDs.

Thank you Charlie.


General

________________________________

**Comment by charliecon** [comment]
I understand now. You're saying it would be better if you could set the ID of your questions and answers yourself when you create the form, so that you can reference them elsewhere in your terraform config?

**Comment by louisdesilva** [comment]
Yes 🙂
Like now, I can reference callflowID in another flow etc.


General

________________________________

**Comment by charliecon** [comment]
We will create a ticket for this and get a developer working on it as soon as possible. The resource will be updated within the next 2/3 weeks. 

If we need any more info from you, or if you have any questions, we can use this thread 🙂

Thanks, Louis!

(Tracking with DEVTOOLING-625)

**Comment by louisdesilva** [comment]
Thanks Charlie. Awesome 🙂


General

________________________________

**Comment by charliecon** [comment]
Hi @louisdesilva 

I tried adding the ID fields to the items in `questions` and `answerOptions` in [POST /api/v2/quality/forms/evaluations](https://developer.genesys.cloud/devapps/api-explorer#post-api-v2-quality-forms-evaluations). The problem I'm facing is that the API generated a random GUID for these IDs regardless of whether I provide a value or not. 

Have you seen this behaviour, or are you able to set your own GUID for these objects? 

I may have to speak to the team that owns the API to see if they can help out with this issue.

-Charlie

**Comment by louisdesilva** [comment]
Hi Charlie,

From what I have seen, the id in the survey form (questionid and answerid) changes each time you make a change and re-publish the form.
and once the form is published, the questionid and answerid GUID remains the same regardless how many times you use it.

Regards,
Louis



General

________________________________

**Comment by charliecon** [comment]
If the API decides what the question/answerOption ID is, it won't make a difference if I add the id field to the resource. I'm talking with the team that owns the API to confirm that this is the case. I'll keep you posted.

**Comment by charliecon** [comment]
Hi @louisdesilva 

Quick update: I added the ID fields to the schemas you mentioned above. They cannot be set by the user as the API will always generate them even if you provide a value, so they will be strictly read-only. Although it would be more convenient to set the IDs yourself and store the value in a local variable, having them as read-only fields means you can reference them in your Terraform config like this:

```hcl
output "question_group_id" {
  value = genesyscloud_quality_forms_evaluation.example.question_groups.0.id
}

output "question_id_1" {
  value  = genesyscloud_quality_forms_evaluation.example.question_groups.0.questions.0.id
}

output "question_id_2" {
  value  = genesyscloud_quality_forms_evaluation.example.question_groups.0.questions.1.id
}
```

All going well, you should see these changes released tomorrow on v1.43.0 

Best regards,
Charlie

**Comment by louisdesilva** [comment]
Thank you Charlie.


General

________________________________
